File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/c/MinJumps.c
#include "java_arrays.h"
#include <stdint.h>

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left + right);
*/
static int32_t java_add(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left + (uint32_t)right);
}

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left - right);
*/
static int32_t java_sub(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left - (uint32_t)right);
}

/*@
  assigns \nothing;
  ensures \result == (left < right ? left : right);
*/
static int32_t java_min(int32_t left, int32_t right) {
    return left < right ? left : right;
}

/*@
  logic integer fb_jumps{L}(JIntArray a, integer i, integer j) =
    i <= 0 ? 0 : j <= 0 ? INT32_MAX : \min(fb_jumps(a, i, j - 1),
    (int32_t)((integer)a->data[j - 1] + j - 1) >= i ? (int32_t)(fb_jumps(a, j - 1, j - 1) + 1) : INT32_MAX);
*/
/*@
  requires n >= 1 && (n == 1 || (jintarray_valid_nonnull(arr) && n - 1 <= arr->length));
  assigns \nothing;
  ensures \result == (n == 1 ? 0 : fb_jumps(arr, n - 1, n - 1));
*/
int32_t minJumps(JIntArray arr, int32_t n) {
    JIntArray dp = jarray_new(n);
    int32_t length = jarray_length(dp);
    /*@
      loop invariant 0 <= i <= length && length == n;
      loop invariant jintarray_valid_nonnull(dp) && dp->length == n;
      loop invariant \forall integer k; 0 <= k < i ==> dp->data[k] == INT32_MAX;
      loop assigns i, dp->data[0 .. n - 1];
      loop variant n - i;
    */
    for (int32_t i = 0; i < length; i = java_add(i, 1)) {
        jarray_set(dp, i, INT32_MAX);
    }
    jarray_set(dp, 0, 0);

    /*@
      loop invariant 1 <= i <= n;
      loop invariant jintarray_valid_nonnull(dp) && dp->length == n;
      loop invariant \forall integer k; 0 <= k < i ==> dp->data[k] == (k == 0 ? 0 : fb_jumps(arr, k, k));
      loop invariant \forall integer k; i <= k < n ==> dp->data[k] == INT32_MAX;
      loop assigns i, dp->data[0 .. n - 1];
      loop variant n - i;
    */
    for (int32_t i = 1; i < n; i = java_add(i, 1)) {
        /*@
          loop invariant 0 <= j <= i;
          loop invariant jintarray_valid_nonnull(dp) && dp->length == n;
          loop invariant dp->data[i] == fb_jumps(arr, i, j);
          loop assigns j, dp->data[i];
          loop variant i - j;
        */
        for (int32_t j = 0; j < i; j = java_add(j, 1)) {
            int32_t reach = java_add(jarray_get(arr, j), j);
            if (reach >= i) {
                int32_t candidate = java_add(jarray_get(dp, j), 1);
                jarray_set(dp, i, java_min(jarray_get(dp, i), candidate));
            }
        }
    }
    return jarray_get(dp, java_sub(n, 1));
}

File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/java/MinJumps.java

import java.io.*;
import java.lang.*;
import java.util.*;
import java.math.*;


class MinJumps {
    
    /*@ public normal_behavior
      @ ensures \result == ((x % 4294967296L + 6442450944L) % 4294967296L) - 2147483648L;
      @ model public static pure \bigint fb_wrap(\bigint x);
      @*/
    /*@ public normal_behavior
      @ ensures \result == (a <= b ? a : b);
      @ model public static pure \bigint fb_min(\bigint a, \bigint b);
      @*/
    /*@ public normal_behavior
      @ requires a != null && 0 <= j <= i <= a.length;
      @ ensures \result == (i == 0 ? 0 : j == 0 ? Integer.MAX_VALUE : fb_min(fb_jumps(a, i, j - 1), \java_math(a[j - 1] + (j - 1)) >= i ? fb_wrap(fb_jumps(a, j - 1, j - 1) + 1) : Integer.MAX_VALUE));
      @ measured_by (\bigint)i * (a.length + 1) + j;
      @ model public static pure \bigint fb_jumps(int[] a, int i, int j);
      @*/
    /*@ public normal_behavior
      @ requires n >= 1 && (n == 1 || (arr != null && n - 1 <= arr.length));
      @ assignable \nothing;
      @ ensures \result == (n == 1 ? 0 : fb_jumps(arr, n - 1, n - 1));
      @*/
    //@ code_java_math
    public static int minJumps(int[] arr, int n) {
        int[] dp = new int[n];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] = 0;

        /*@ loop_invariant 1 <= i <= n;
          @ loop_invariant (\forall int k; 0 <= k < i; dp[k] == (k == 0 ? 0 : fb_jumps(arr, k, k)));
          @ loop_invariant (\forall int k; i <= k < n; dp[k] == Integer.MAX_VALUE);
          @ loop_writes i, dp[*];
          @ decreases n - i;
          @*/
        for (int i = 1; i < n; i++) {
            /*@ loop_invariant 0 <= j <= i;
              @ loop_invariant dp[i] == fb_jumps(arr, i, j);
              @ loop_writes j, dp[i];
              @ decreases i - j;
              @*/
            for (int j = 0; j < i; j++) {
                if (arr[j] + j >= i) {
                    dp[i] = Math.min(dp[i], dp[j] + 1);
                }
            }
        }
        return dp[n - 1];
    }
}

